Hybrid bootloader (BIOS+UEFI) installation support

ArrayBolt3 opened the pull request:
It would be very handy for some use case scenarios if Calamares could be configured to install both BIOS and GRUB bootloaders on the same system. The primary use for this would be portable installations, where someone does a full install of a distro onto a USB flash drive that they then intend to run on multiple physical machines with varying types of firmware. Currently Calamares can only install the bootloader matching the firmware type of the system being booted though. This PR attempts to change that by adding an option `installHybridGRUB` to `bootloader.conf`. When enabled, Calamares will attempt to install both BIOS and UEFI GRUB bootloaders.

Note, I'm not sure if the approach I'm taking here is correct - it seems a bit overly hacky. This PR also isn't complete - it still needs to tackle a problem with partitioning (namely, the EFI system partition is only made on EFI systems, so on BIOS systems, one has to specify an ESP themselves in partition.conf... but if you do that, you end up with two ESPs on EFI systems, because Calamares will add an ESP regardless of whether one already exists in the partition layout or not).

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,25 +637,47 @@ def run_grub_install(fw_type, partitions, efi_directory):
+            # drive the ESP is located on.
+            installation_root_path = libcalamares.globalstorage.value("rootMountPoint")
+            install_efi_directory = installation_root_path + efi_directory
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source" \
+                + install_efi_directory \
+                + " | tail -n1) | tail -n1)\""
```
Would this work(and be simpler)?  `f"findmnt -no source {install_efi_directory}"`

Also, try to use f strings when it can improve readability.  When the module was written it wasn't supported but for new code it is available now.

That being said, I am not sure this should be really be happening here.  It seems like this should come from the partition module.

ArrayBolt3 replied:
`findmnt` doesn't quite work because it gives me the partition device file, not the disk device file (which is what `grub-install` wants when doing a BIOS-style installation). It could probably replace the `df` stuff in the innermost subshell, but the `lsblk` and `echo` will still be needed if this kind of logic is used.

dalto8 replied:
Is that not in global storage already in `partitions`?

ArrayBolt3 replied:
It does not seem to be based on some debugging. On BIOS systems, it's in the bootLoader key, but on EFI systems it's not there at all. On the VM I just tested this on, there's some strange "empty" device (`{'claimed': False, 'device': '', 'features': {}, 'fs': 'unknown', 'fsName': 'unknown', 'mountPoint': '', 'partattrs': 0, 'partlabel': '', 'parttype': '', 'partuuid': '', 'uuid': ''}`), then there's four devices, one for each of /dev/vda1 through /dev/vda4. /dev/vda itself isn't in the list.

I don't want to just transform the partition name to a disk name since it's not guaranteed the two will have a clear correlation  - we see a minor version of this with /dev/sda and /dev/sda1 vs. /dev/mmcblk0 and /dev/mmcblk0p1.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,25 +637,47 @@ def run_grub_install(fw_type, partitions, efi_directory):
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source" \
+                + install_efi_directory \
+                + " | tail -n1) | tail -n1)\""
+            boot_loader_install_path = subprocess.run(["sh", "-c", find_esp_disk_command],
+                capture_output=True).stdout.decode().strip()
```
Please don't use subprocess.run, instead use Calamares's python functions for running commands.

ArrayBolt3 replied:
Got it, latest commits do it the correct way.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -923,7 +956,11 @@ def run():
 
     try:
-        prepare_bootloader(fw_type)
+        if (boot_loader == "grub" or efi_boot_loader == "grub") and install_hybrid_grub:
+            prepare_bootloader("bios")
+            prepare_bootloader("efi")
+        else:
+            prepare_bootloader(fw_type)
```
I think this would be cleaner if we implemented as part of the bootloader selection logic inside `prepare_bootloader()`

ArrayBolt3 replied:
Moved.

dalto8 reviewed (commented):
Specific comments below but, generally, it seems like this really needs changes and support in the partition module.  This will otherwise be too fragile I think.

We need to be ensuring that the partition setup is correct for this implementation method.  It would also solve the issue with the duplicate ESPs.

ArrayBolt3 commented:
Agreed, the partition module will need some help. I'd really like if there's some way to make it so that this change only requires things to be tweaked in one spot (a bootloader setting) rather than needing both bootloader and partition config changes, but I'm not sure how practical that is yet.

Thanks for the feedback, I'll try some more things and see if I can improve on this. (This implementation actually does work at least in the most straightforward scenarios, if you combine it with some partition layout tweaking, but I'm not thrilled with it and would like to make it cleaner.)

ArrayBolt3 commented:
It's definitely not perfect yet, but this is significantly better than the initial draft. Support for getting the partition layout right regardless of host firmware is now added to the partition module, and the actual bootloader installation code is a bit less messy. Currently the BIOS boot partition is simply labeled "New Partition", so that will need fixed, but this is getting close to ready.

ArrayBolt3 commented:
@dalto8 Would like another preliminary review when you have time. This is pretty close to ready, I have BIOS boot partition detection "working", though I'm not satisfied with the strategy I'm using at the moment (it's assuming any filesystemless partition 8 MiB or less in size is BIOS boot, which may not always be true, but checking for a bios-boot flag turns out to be hard).

ArrayBolt3 commented:
Alright, I fixed the bios-grub partition detection problem! I think this is now ready for review. Probably will need a commit squash before merge though.

ArrayBolt3 committed:
[bootloader] Add hybrid GRUB installation support

Some users may want to install their system in a way that it can be
booted with both BIOS and UEFI firmware. This requires installing both
BIOS and UEFI bootloaders. This commit adds the ability to do this
with GRUB.

ArrayBolt3 committed:
[partition] Add hybrid bootloader compatible partitioning mode

Some users may want to install their system in a way that it can be
booted with both BIOS and UEFI firmware. This requires installing both
BIOS and UEFI bootloaders. This commit adds the ability to
automatically partition a disk in a way that is compatible with this
setup.

dalto8 commented on src/modules/partition/core/PartitionActions.cpp:
```diff
@@ -105,7 +110,11 @@ doAutopartition( PartitionCoreModule* core, Device* dev, Choices::AutoPartitionO
+        if ( isEfi || createHybridBootloaderLayout )
+        {
+            partType = PartitionTable::gpt;
+        }
```
```suggestion
        partType = ( isEfi || createHybridBootloaderLayout ) ? PartitionTable::gpt : PartitionTable::msdos;
```

ArrayBolt3 replied:
Applied.

ArrayBolt3 commented:
Alright, commits now cleaned up a bit.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -909,8 +914,19 @@ def run():
     fw_type = libcalamares.globalstorage.value("firmwareType")
+    boot_loader = libcalamares.globalstorage.value("bootLoader")
+
+    try:
+        install_hybrid_grub = libcalamares.job.configuration["installHybridGRUB"]
+    except KeyError:
+        install_hybrid_grub = False
```
```suggestion
    install_hybrid_grub = libcalamares.job.configuration.get("installHybridGRUB", False)
```

ArrayBolt3 replied:
Applied.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -909,8 +914,19 @@ def run():
+    except KeyError:
+        install_hybrid_grub = False
+
+    try:
+        efi_boot_loader = libcalamares.job.configuration["efiBootLoader"]
+    except KeyError:
+        efi_boot_loader = ""
```
```suggestion
    efi_boot_loader = libcalamares.job.configuration.get("efiBootLoader", "")
```

ArrayBolt3 replied:
Applied.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -923,7 +939,12 @@ def run():
     try:
-        prepare_bootloader(fw_type)
+        prepare_bootloader(fw_type, install_hybrid_grub)
+#        if efi_boot_loader == "grub" and install_hybrid_grub:
+#            prepare_bootloader("bios")
+#            prepare_bootloader("efi")
+#        else:
+#            prepare_bootloader(fw_type)
```
If this isn't needed we should remove it.

ArrayBolt3 replied:
Removed.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -893,8 +898,8 @@ def prepare_bootloader(fw_type):
         install_refind(efi_directory)
-    elif efi_boot_loader == "grub" or fw_type != "efi":
-        install_grub(efi_directory, fw_type)
+    elif efi_boot_loader == "grub" or fw_type != "efi" or install_hybrid_grub:
```
Are we sure we want this check here?  Do we really want to follow the hybrid grub path on EFI system where something other than "grub" is specified?

ArrayBolt3 replied:
Hmm, that's a good question. Really, if `install_hybrid_grub` is `True`, setting a boot loader other than GRUB should be illegal. That would make this check obsolete, since the BIOS path will always install with GRUB anyway, and the EFI path will install with GRUB if it's required in order to use `install_hybrid_grub`.

ArrayBolt3 replied:
Well this is fine, looks like `bootloader.schema.yaml` is being ignored entirely. Not sure why, but I can set any parameter to any value and Calamares doesn't choke on it when first loading.

ArrayBolt3 replied:
@dalto8 Are the schema.yaml files actually used to check config at Calamares startup? I just tried using git master, 3.3.14, 3.3.5 (I think), and 3.2.61 and *all* of them are unable to notice when I mangle a module's config file.

ArrayBolt3 replied:
Gave up on using the schema, added a validation blurb near the start of `run()` and removed the unnecessary check.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,28 +637,33 @@ def run_grub_install(fw_type, partitions, efi_directory):
 
 
-def install_grub(efi_directory, fw_type):
+def install_grub(efi_directory, fw_type, install_hybrid_grub):
```
Don't forget to add the new param to the doc

ArrayBolt3 replied:
Added.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -671,7 +676,7 @@ def install_grub(efi_directory, fw_type):
 
-    if fw_type == "efi":
+    if fw_type == "efi" or install_hybrid_grub:
         libcalamares.utils.debug("Bootloader: grub (efi)")
```
Should we add the state of `install_hybrid_grub` to this debug message?

ArrayBolt3 replied:
Yeah, that would be a good idea.

ArrayBolt3 replied:
Added.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -712,9 +717,9 @@ def install_grub(efi_directory, fw_type):
         libcalamares.utils.debug("Bootloader: grub (bios)")
-        run_grub_install(fw_type, partitions, None)
+        run_grub_install("bios", partitions, efi_directory, install_hybrid_grub)
 
```
We should probably have an error case here in the event we fall through all those conditions.

ArrayBolt3 replied:
`fw_type` is essentially a boolean right now - it's only ever `bios` or `efi`, to my awareness. However, I do see how it could end up blank if something goes wrong, and Calamares might in theory support more firmware types in the future. Maybe it would be a good idea to error out at the *top* if `fw_type` is anything other than `bios` or `efi`? That would be more efficient and still catch the kinds of problems that could surface here.

ArrayBolt3 replied:
Added a check at the top as described.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,28 +637,33 @@ def run_grub_install(fw_type, partitions, efi_directory):
-        boot_loader = libcalamares.globalstorage.value("bootLoader")
-        if boot_loader["installPath"] is None:
-            return
+        if libcalamares.globalstorage.value("bootLoader") is None:
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source /boot/efi | tail -n1) | tail -n1)\""
+            boot_loader_install_path = check_target_env_output(["sh", "-c", find_esp_disk_command]).strip()
```
If bootloader is None and fw_type is efi and install_hybrid_grub is false shouldn't we be not installing a bootloader here?

ArrayBolt3 replied:
Yes, but I think that's already the case, since the context is:

```python
if fw_type == "efi":
    ...do EFI things here...
else:
   ...
   libcalamares.globalstorage.value("bootLoader") is None:
        find_esp_disk_command = "long line of bash"
        boot_loader_install_path = check_target_env_output([...]).strip()
```

i.e., this code will only run if `fw_type` is NOT `efi`.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,28 +637,33 @@ def run_grub_install(fw_type, partitions, efi_directory):
-        if boot_loader["installPath"] is None:
-            return
+        if libcalamares.globalstorage.value("bootLoader") is None:
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source /boot/efi | tail -n1) | tail -n1)\""
```
We shouldn't hardcode `/boot/efi`

The efi mountpoint is configurable in the partition module.

ArrayBolt3 replied:
gah, I knew that, but forgot it here. Will fix in the next push.

ArrayBolt3 replied:
Um, wait, the efi mountpoint *is* configurable, but I'm pretty sure that's stored in `libcalamares.globalstorage.value("bootLoader")["installPath"]` no? This is the codepath for if `libcalamares.globalstorage.value("bootLoader") is None`, in which case we don't have that info. Is there another convenient spot to grab it from, or should I be parsing the partition layout to find the right spot and falling back to `/boot/efi` if all else fails?

dalto8 replied:
I think it is also in global storage.  Something like `efiSystemPartition`

ArrayBolt3 replied:
Was able to use `efiSystemPartition`. `/boot/efi` is still used as a final fallback, but only if all other options fail.

dalto8 commented on src/modules/bootloader/main.py:
```diff
@@ -637,28 +637,33 @@ def run_grub_install(fw_type, partitions, efi_directory):
-            return
+        if libcalamares.globalstorage.value("bootLoader") is None:
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source /boot/efi | tail -n1) | tail -n1)\""
+            boot_loader_install_path = check_target_env_output(["sh", "-c", find_esp_disk_command]).strip()
```
We should probably check to ensure that `boot_loader_install_path` is something like we expect.

ArrayBolt3 replied:
Added a check to make sure the path begins with `/dev/` and is at least eight characters long (the length of `/dev/sda`, to my awareness there aren't any two-character disk names in Linux).

ArrayBolt3 commented:
Did a smoke test (hybrid installation on BIOS hardware, try to boot it with both BIOS and UEFI hardware), the latest commits appear to work.

adriaandegroot commented on src/modules/bootloader/main.py:
```diff
@@ -637,42 +637,61 @@ def run_grub_install(fw_type, partitions, efi_directory):
-        assert efi_directory is None
-        if libcalamares.globalstorage.value("bootLoader") is None:
-            return
+        if efi_directory is not None:
```
Perhaps combine these? `if efi-directory-set but not install-hybrid: print(error); return;`, I think it would be good to log some kind of error anyway.

ArrayBolt3 replied:
Good idea. Implemented a warning message with `libcalamares.utils.warning`.

adriaandegroot commented on src/modules/bootloader/main.py:
```diff
@@ -637,42 +637,61 @@ def run_grub_install(fw_type, partitions, efi_directory):
+            efi_install_path = libcalamares.globalstorage.value("efiSystemPartition")
+            if efi_install_path is None or efi_install_path == "":
+                efi_install_path = "/boot/efi"
+            find_esp_disk_command = f"echo \"/dev/$(lsblk -o PKNAME $(df --output=source '{efi_install_path}' | tail -n1) | tail -n1)\""
```
I get the feeling we're doing things in shell that we don't have to. `lsblk $(df | tail -n1)` ok, you want the substitution there, and *maybe* the second `tail -n1` as well, but why do string-manipulation there as well, .. you could do this in one or more steps with target_env_output and then add the `/dev/` bit in Python.

ArrayBolt3 replied:
Valid. You may be able to tell I'm a bit handier with bash than with Python by this :P

The second `tail -n1` can be removed without much trouble, and it removes a fork, which should help performance a tiny bit.

adriaandegroot commented on src/modules/partition/core/PartitionActions.cpp:
```diff
@@ -144,6 +149,26 @@ doAutopartition( PartitionCoreModule* core, Device* dev, Choices::AutoPartitionO
+        if ( createHybridBootloaderLayout )
+        {
+            // TODO: Avoid hardcoding 8 MiB?
+            qint64 bios_part_sizeB = 8388608;
```
537:    qint64 uefisys_part_sizeB = 300_MiB;

ArrayBolt3 replied:
Oh nice, didn't know that we had `_MiB` literals in Cala. I see them now. Will use this.

adriaandegroot commented on src/modules/partition/gui/PartitionLabelsView.cpp:
```diff
@@ -190,6 +192,11 @@ PartitionLabelsView::buildTexts( const QModelIndex& index ) const
                 firstLine = tr( "EFI system", "@label" );
             }
+            else if ( index.data( PartitionModel::FileSystemTypeRole ).toInt() == FileSystem::Unformatted
+                      && static_cast< Partition* >( index.data( PartitionModel::PartitionPtrRole ).value< void* >() )->property( "_calamares_flags" ).toUInt() & KPM_PARTITION_FLAG( BiosGrub ) )
```
Can we give all that casting-and-wrangling a nicer name, make it a static function in this translation unit?

ArrayBolt3 replied:
Will do.

ArrayBolt3 commented:
Gentle ping, anything left to do on this?
